core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageInputStream.java

```diff
@@ -0,0 +1,171 @@
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class StreamMessageInputStream<T> extends InputStream {
+
+    private final StreamMessage<T> source;
+    private final EventExecutor executor;
+    private final StreamMessageInputStreamSubscriber<T> subscriber;
+    private final AtomicBoolean subscribed = new AtomicBoolean();
+    private volatile boolean closed;
+
+    StreamMessageInputStream(StreamMessage<T> source,
+                             Function<? super T, ? extends HttpData> httpDataConverter,
+                             EventExecutor executor) {
+        requireNonNull(source, "source");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        requireNonNull(executor, "executor");
+        this.source = source;
+        this.executor = executor;
+        subscriber = new StreamMessageInputStreamSubscriber<>(httpDataConverter);
+    }
+
+    @Override
+    public int read() throws IOException {
+        return read(in -> {
+            try {
+                return in.read();
+            } catch (IOException e) {
+                return Exceptions.throwUnsafely(e);
+            }
+        });
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        return read(in -> {
+            try {
+                return in.read(b, off, len);
+            } catch (IOException e) {
+                return Exceptions.throwUnsafely(e);
+            }
+        });
+    }
+
+    private int read(Function<InputStream, Integer> function) throws IOException {
+        ensureOpen();
+        ensureSubscribed();
+        subscriber.request();
+        return function.apply(byteBufsInputStream());
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (closed) {
+            return;
+        }
+        closed = true;
+        source.abort();
+        byteBufsInputStream().close();
+    }
+
+    @Override
+    public int available() throws IOException {
+        ensureOpen();
+        return byteBufsInputStream().available();
+    }
+
+    private void ensureOpen() throws IOException {
+        if (closed) {
+            throw new IOException("Stream closed");
+        }
+    }
+
+    private void ensureSubscribed() {
+        if (subscribed.compareAndSet(false, true)) {
+            source.subscribe(subscriber, executor);
+        }
+        subscriber.whenSubscribed.join();
+    }
+
+    private ByteBufsInputStream byteBufsInputStream() {
+        return subscriber.byteBufsInputStream;
+    }
+
+    private static final class StreamMessageInputStreamSubscriber<T> implements Subscriber<T> {
+
+        private final Function<? super T, ? extends HttpData> httpDataConverter;
+        @Nullable
+        private volatile Subscription upstream;
+        private final CompletableFuture<Void> whenSubscribed = new CompletableFuture<>();
+        private final ByteBufsInputStream byteBufsInputStream = new ByteBufsInputStream();
```

Comment:
nit: Could we organize `final` fields together?

Reply:
✅ updated.